csharp_dotnetcore/11.QnAMaker/QnAMaker/BotServices.cs

```diff
@@ -0,0 +1,51 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License
+
+using System;
+using System.Collections.Generic;
+using Microsoft.ApplicationInsights;
+using Microsoft.Bot.Builder.AI.QnA;
+
+namespace AspNetCore_QnA_Bot
+{
+    /// <summary>
+    /// Class represents the bot's references to external services.
+    ///
+    /// For example, Application Insights and QnaMaker services
+    /// are kept here (singletons).  These external services are configured
+    /// using the BotConfigure class (based on the contents of your ".bot" file).
+    /// </summary>
+    [Serializable]
+    public class BotServices
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="BotServices"/> class.
+        /// </summary>
+        public BotServices()
+        {
+            // Given there can be multiple Qna services used in a single bot,
+            // Qna is represented as a Dictionary.  This is also modeled in the
+            // ".bot" file since the elements are named (string).
+            QnAServices = new Dictionary<string, QnAMaker>();
+        }
+
+        /// <summary>
+        /// Gets or sets the Application Insights Telemetry client.
+        /// Use this to log new custom events/metrics/traces/etc into your
+        /// Application Insights service for later analysis.
+        /// </summary>
+        /// <value>
+        /// The Application Insights Telemetry client instance created based on configuration in the .bot file.
+        /// </value>
+        public TelemetryClient TelemetryClient { get; set; }
+
+        /// <summary>
+        /// Gets or sets the (potential) set of Qna Services used.
+        /// This sample only uses a single Qna instance.
+        /// </summary>
+        /// <value>
+        /// A QnaMaker client instance created based on configuration in the .bot file.
+        /// </value>
+        public Dictionary<string, QnAMaker> QnAServices { get; set; }
```

Comment:
public Dictionary<string, QnAMaker> QnAServices { get; set; } = new Dictionary<string, QnAMaker>();

Reply:
In what instances would we want this to be set? I can envision *clearing* the dictionary, but replacing the dictionary itself seems like something we don't need. 

Consider making this a "get" only property.

Reply:
I believe allowing the dictionary to be replaced opens up additional concerns around concurrency, as to even look at the dictionary requires a lock, as a different ASP.Net thread could be changing the property to a different dictionary instance. Easier to initialize during object construction and never allow it to be replaced.